Everywhere: Remove cyclic #include-s

BenWiederhake opened the pull request:
These happen to be non-problematic thanks to sheer luck, but let's avoid trouble down the road.

A tool for detecting these automatically during pre-commit is coming. :^)

trflynn89 commented on Userland/Libraries/LibDebug/Dwarf/DwarfInfo.h:
```diff
@@ -7,14 +7,15 @@
 #include <AK/RedBlackTree.h>
 #include <AK/RefCounted.h>
 #include <AK/String.h>
+#include <LibDebug/Dwarf/CompilationUnit.h>
+#include <LibDebug/Dwarf/DIE.h>
```
These don't seem needed to be included in the header, as their types are used only as pointers & template params?

BenWiederhake replied:
They are needed. The compiler says so, and for a good reason. DwarfInfo contains
```
    NonnullOwnPtrVector<Dwarf::CompilationUnit> m_compilation_units;
```
and an implicit destructor, which needs to be able to call the destructor of CompilationUnit. The latter is impossible with an incomplete definition of CompilationUnit.

You're right about DIE.h being unnecessary; it is already included by CompilationUnit.

trflynn89 replied:
> You're right about DIE.h being unnecessary; it is already included by CompilationUnit.

I wouldn't remove it for that reason, implicit includes are not good.

IMO a out-of-line destructor is the way to go here as well.

BenWiederhake replied:
Okay, I'll keep the include.

Why put the destructor out of line? The cyclic includes are already fixed by LineProgram no longer including DwarfInfo.

If you want to remove the CompilationUnit include, then that will probably break `DwarfInfo::for_each_compilation_unit` in weird ways.

trflynn89 commented on Userland/Libraries/LibWeb/CSS/Parser/DeclarationOrAtRule.h:
```diff
@@ -7,6 +7,7 @@
 #pragma once
 
 #include <LibWeb/CSS/Parser/Declaration.h>
+#include <LibWeb/CSS/Parser/Function.h>
```
Why's this one needed? `Function` doesn't appear in this header

BenWiederhake replied:
You're right, thanks! I intended to add it to the `.cpp` file.

In case you're curious: DeclarationOrAtRule contains a Declaration contains a ComponentValue contains a Function. So when the destructor is defined in `DeclarationOrAtRule.cpp`, the full definition of `Function` is necessary.

BenWiederhake commented on Userland/Libraries/LibDebug/Dwarf/CompilationUnit.h:
```diff
@@ -15,8 +15,6 @@
 
 class DwarfInfo;
-class DIE;
-class LineProgram;
```
It can't, because CompilationUnit contains
```
NonnullOwnPtr<LineProgram> m_line_program;
```
and an implicit destructor, which needs to be able to call the destructor of LineProgram. The latter is impossible with an incomplete definition of LineProgram.

trflynn89 replied:
Ah, the common workaround for that is to add an out-of-line `= default` destructor in the .cpp file (which IMO is preferable over not being able to forward declare)

BenWiederhake replied:
I can do that, sure, but that sounds like extra pain, and also `line_program()` must move out of line. What for? Breaking the cycle at LineProgram is sufficient, it doesn't need to be broken again.

trflynn89 replied:
Reducing `#include`s helps build times (a *lot*, over a whole codebase) and prevents further opportunity for cycles. I'd argue that's less painful than writing `CompilationUnit::CompilationUnit() = default` in a .cpp file.

Anyways, it's more of a "while we're here doing the right thing, let's do more of the right thing" point, but I'm not gonna hold up this PR for it.

BenWiederhake replied:
Yeah, good point, you're right. I'll add it. :+1:

BenWiederhake commented:
Changes since last push:
- Rebased for convenience
- DwarfInfo.h now forward-declares CompilationUnit instead of including it, as suggested by @trflynn89 
- CompilationUnit.h now forward-declares LineProgram instead of including it, as suggested by @trflynn89

I don't expect this to actually reduce compile times, but it definitely doesn't hurt.

trflynn89 commented on Userland/Libraries/LibWeb/Layout/InitialContainingBlock.h:
```diff
@@ -8,6 +8,7 @@
 
 #include <LibWeb/DOM/Document.h>
 #include <LibWeb/Layout/BlockContainer.h>
+#include <LibWeb/Layout/LayoutPosition.h>
```
Not sure why this include is needed in the .h? It's only used as a `const&` here, does it suffice to be in the .cpp instead?

BenWiederhake replied:
It includes Document.h which includes ParentNode.h which includes Node.h which used to include LayoutPosition.h. This used to be the only path through which `Layout::LayoutPosition` got declared. Perhaps a declaration is sufficient, but it definitely needs to be declared, at minimum.

It also needs a definition of `Layout::LayoutRange`, so that the method `LayoutRange const& InitialContainingBlock::selection() const` knows how to arrange the registers. `Layout::LayoutRange` is defined in `LayoutPosition.h`, so we need to include that header.

In theory you could split that header into two, but it is very short (i.e. cannot significantly shorten the compile time, it's barely even 60 lines!), and was included anyway (i.e. this PR does not increase the number of included headers). You are very welcome to shorten compile times! I'd like to do something different in this PR, namely remove cyclic includes.

trflynn89 replied:
My dude, I asked just because it is not obvious why a header is needed when a forward declaration would surely suffice. It's a mistake that this type isn't in Forward.h already, but yes that can be another PR.

BenWiederhake replied:
:+1: Great!

I don't think that a forward declaration will suffice, due to `LayoutRange`, but I'm looking forward to your PR :)

bgianfo commented on Userland/Libraries/LibCpp/Parser.h:
```diff
@@ -6,12 +6,12 @@
 #pragma once
 
-#include "AK/NonnullRefPtr.h"
 #include "AST.h"
```
Why not `AST.h` as well?

BenWiederhake replied:
AST.h is in the same directory, so it's not a "badly-formatted include". But yeah, good idea, I'll change it.

bgianfo replied:
Sure, you just updated others in the same directory, so I was confused why that was special. 😁

BenWiederhake replied:
It must have slipped my attention ^^

BenWiederhake commented:
Changes since last push:
- Rebased for convenience
- Changed the local include in `LibCpp/Parser.h` to a system include (thanks bgianfo!)
- Triple-checked that this does not cause any miscompiling headers (#15231)
